Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp

```diff
@@ -99,65 +100,146 @@ namespace Terrain
+            worldBounds, &AzFramework::Terrain::TerrainDataRequests::GetTerrainAabb);
+
+        const AZ::Aabb& regionToUpdate = dirtyRegion.IsValid() ? dirtyRegion : worldBounds;
+
+        m_dirtyRegion.AddAabb(regionToUpdate);
+        m_dirtyRegion.Clamp(worldBounds);
+
+        AZ::Transform transform = AZ::Transform::CreateTranslation(worldBounds.GetCenter());
+        
+        AZ::Vector2 queryResolution = AZ::Vector2(1.0f);
+        AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
+            queryResolution, &AzFramework::Terrain::TerrainDataRequests::GetTerrainHeightQueryResolution);
+
+        uint32_t width = aznumeric_cast<uint32_t>((float)m_dirtyRegion.GetXExtent() / queryResolution.GetX());
+        uint32_t height = aznumeric_cast<uint32_t>((float)m_dirtyRegion.GetYExtent() / queryResolution.GetY());
+        
+        uint32_t worldWidth = aznumeric_cast<uint32_t>((float)worldBounds.GetXExtent() / queryResolution.GetX());
+        uint32_t worldHeight = aznumeric_cast<uint32_t>((float)worldBounds.GetYExtent() / queryResolution.GetY());
+
         m_areaData.m_transform = transform;
         m_areaData.m_heightScale = worldBounds.GetZExtent();
         m_areaData.m_terrainBounds = worldBounds;
-        m_areaData.m_heightmapImageHeight = height;
-        m_areaData.m_heightmapImageWidth = width;
-        m_areaData.m_sampleSpacing = sampleSpacing;
+        m_areaData.m_heightmapImageWidth = worldWidth;
+        m_areaData.m_heightmapImageHeight = worldHeight;
+        m_areaData.m_updateWidth = width;
+        m_areaData.m_updateHeight = height;
+        m_areaData.m_sampleSpacing = queryResolution.GetX();
+        m_areaData.m_propertiesDirty = true;
+    }
+
+    void TerrainFeatureProcessor::UpdateTerrainData()
+    {
+        uint32_t width = m_areaData.m_updateWidth;
+        uint32_t height = m_areaData.m_updateHeight;
+        const AZ::Aabb& worldBounds = m_areaData.m_terrainBounds;
+        float queryResolution = m_areaData.m_sampleSpacing;
+
+        AZ::RHI::Size worldSize = AZ::RHI::Size(m_areaData.m_heightmapImageWidth, m_areaData.m_heightmapImageHeight, 1);
+
+        if (!m_areaData.m_heightmapImage || m_areaData.m_heightmapImage->GetDescriptor().m_size != worldSize)
+        {
+            // World size changed, so the whole world needs updating.
+            width = worldSize.m_width;
+            height = worldSize.m_height;
+            m_dirtyRegion = worldBounds;
+
+            AZ::Data::Instance<AZ::RPI::AttachmentImagePool> imagePool = AZ::RPI::ImageSystemInterface::Get()->GetSystemAttachmentPool();
+            AZ::RHI::ImageDescriptor imageDescriptor = AZ::RHI::ImageDescriptor::Create2D(
+                AZ::RHI::ImageBindFlags::ShaderRead, width, height, AZ::RHI::Format::R16_UNORM
+            );
+            m_areaData.m_heightmapImage = AZ::RPI::AttachmentImage::Create(*imagePool.get(), imageDescriptor, AZ::Name("TerrainHeightmap"), nullptr, nullptr);
```

Comment:
The AZ::Name("TerrainHeightmap") seems like it could be a const member of the FP...

Reply:
good call